### ISSUE TITLE
Setting values in arguments still flags `unused var`.

### ISSUE THREAD

**Issue by prograhammer** [eslint/eslint#8223]
**Tell us about your environment**
ESLint Version: 3.17.1  
Node Version: 6.2.2  
npm Version: 3.9.5  

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint 6.1.2

**Configuration:**
```
module.exports = {
  root: true,
  parser: 'babel-eslint',
  parserOptions: {
    sourceType: 'module'
  },
  // https://github.com/feross/standard/blob/master/RULES.md#javascript-standard-style
  extends: 'vue',
  // required to lint *.vue files
  plugins: [
    'vue',
    'html'
  ],
  env: {
    browser: true
  },
  // add your custom rules here
  'rules': {
    // allow paren-less arrow functions
    'arrow-parens': 0,
    // allow async-await
    'generator-star-spacing': 0,
    // allow debugger during development
    'no-debugger': process.env.NODE_ENV === 'production' ? 2 : 0,
    'no-return-assign': 2,
    // disallow indentation using both tabs and spaces
    'no-mixed-spaces-and-tabs': 2,
    // ensure consistent 2 space indentation and indent cases under switch
    'indent': [2, 2, {'SwitchCase': 1}]
  }
}
```
**What did you do? Please include the actual source code causing the issue.**  

I get `no-unused-vars` for assignments in function arguments:  
```
let bar
this.foo(bar = false)
```

**Comment by kaicataldo** [comment]
I'm assuming this is a simplified case, but in your example `bar` is assigned to but is never used and the rule is actually working as intended.

**Comment by kaicataldo** [comment]
Is the actual case throwing an error slightly different?

**Comment by vitorbal** [comment]
ping @prograhammer. Can you take a look at the questions above when you have a moment?

**Comment by prograhammer** [comment]
Sorry guys for late response. `bar` is passed as argument to a function after assigned false in last line of that code snippet. So it is being used.

**Comment by not-an-aardvark** [comment]
Could you please provide a more complete example? In the code you've given us it seems like the rule is working correctly, so it's hard to determine whether this is a bug.

**Comment by prograhammer** [comment]
No that's the code I tried. Am I misunderstanding the rule? In the snippet is bar still considered unused and why?

**Comment by platinumazure** [comment]
@not-an-aardvark I'm not sure the rule is working correctly.

This is lint-free:

```js
let bar
bar = false
this.foo(bar)
```

This is not, despite being semantically equivalent:

```js
let bar
this.foo(bar = false)
```

I think we need to look at AssignmentExpressions and see if they are themselves used (e.g., by being in an argument of a function call).

**Comment by not-an-aardvark** [comment]
Yes, it's still considered unused, because the value of `bar` is never read. So the code can be replaced with:

```js
this.foo(false)
```

**Comment by vitorbal** [comment]
@prograhammer I think @not-an-aardvark's comment above explains well why the variable is considered unused in your example. Does that make sense?

**Comment by alberto** [comment]
Closing this issue as it looks like the question has been answered. Please feel free to visit us in the ESLint Gitter if you have any other issues!
